Introduce `stack run` command line option

rszibele opened the pull request:
This pull request introduces `stack run` which allows directly running the first available executable in the project similar to `cabal run`. This implements the whishlist issue: https://github.com/commercialhaskell/stack/issues/233

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [ ] The documentation has been updated, if necessary.

I have manually verified that it does indeed run the first available executable and have run the following manual tests:
* `stack exec stack run`
* `stack exec -- stack run help`
* `stack exec -- stack run -- help`

Any thoughts on the code? improvements? I'm pretty new to Haskell and stack.

AshleyYakeley commented:
Does this build first?

rszibele commented:
@AshleyYakeley No, just like with `stack exec` the `stack run` command requires the project to already be built.

AshleyYakeley commented:
OK, so this does _not_ implement #233, which is a `stack run` command that builds and then runs.

AshleyYakeley commented:
OK, so will it build the entire project, or just enough to build the executable in question?

rszibele commented:
I'm tending towards only building the executable that should be run. Given that the dependencies are properly defined, it shouldn't be required to build everything, at least that's my line of thinking. Thoughts?

AshleyYakeley commented:
I think that's sensible.

rszibele commented:
`stack run` now also builds the executable prior to running it, similar to `stack build :myexecutable`. It now also supports specifying an executable as the first command and if it is omitted, it will default to the first available executable in the project.

I have edited my pull request to reflect these changes.

snoyberg commented on src/main/Main.hs:
```diff
@@ -845,6 +853,21 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
+      getRunCmd args = do
+          pkgComponents <- liftM (map lpvComponents . Map.elems . lpProject) getLocalPackages
+          let executables = filter isCExe $ concatMap Set.toList pkgComponents
+          let argExe = if not (null args) then find (\x -> x == (CExe $ T.pack $ head args)) executables
```
I realize that the `not (null args)` ensures that `head args` is total. However, I still prefer to avoid partial function calls, since they can easily turn into bugs with a refactor. Would it be possible to replace this with a pattern match, e.g.:

```haskell
case args of
  [] -> Nothing
  firstArg:_ -> ...
```

snoyberg commented on src/main/Main.hs:
```diff
@@ -845,6 +853,21 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
+          let executables = filter isCExe $ concatMap Set.toList pkgComponents
+          let argExe = if not (null args) then find (\x -> x == (CExe $ T.pack $ head args)) executables
+                                          else Nothing
+          let firstExe = if not (null executables) then Just $ head executables else Nothing
```
Same applies here, though I think you can also use `listToMaybe`.

snoyberg commented on src/main/Main.hs:
```diff
@@ -845,6 +853,21 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
+          let argExe = if not (null args) then find (\x -> x == (CExe $ T.pack $ head args)) executables
+                                          else Nothing
+          let firstExe = if not (null executables) then Just $ head executables else Nothing
+          let (exe, args') = if isJust argExe then (argExe, tail args) else (firstExe, args)
```
And this could be combined with the pattern match in the above `let` binding to avoid the partial `tail` call.

snoyberg commented on src/main/Main.hs:
```diff
@@ -845,6 +853,21 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
+          let (exe, args') = if isJust argExe then (argExe, tail args) else (firstExe, args)
+          case exe of
+              Just (CExe exe') -> do
+                Stack.Build.build (const (return ())) Nothing defaultBuildOptsCLI{boptsCLITargets = [T.concat [T.pack ":", exe']]}
```
Slightly more idiomatic may be: `T.cons ':' exe'`.

snoyberg commented on src/main/Main.hs:
```diff
@@ -845,6 +853,21 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
+                Stack.Build.build (const (return ())) Nothing defaultBuildOptsCLI{boptsCLITargets = [T.concat [T.pack ":", exe']]}
+                return (T.unpack exe', args')
+              _                -> liftIO $ do
+                  hPutStrLn stderr "No executables found."
```
This should probably use `logError` instead, which would need to be outside of `liftIO`.

snoyberg commented on src/main/Main.hs:
```diff
@@ -786,6 +792,7 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
         ExecOptsPlain -> do
           (cmd, args) <- case (eoCmd, eoArgs) of
                 (ExecCmd cmd, args) -> return (cmd, args)
+                (ExecRun, args) -> return ("", args)
```
Can you explain why we need an empty command case here? Is this a code path which shouldn't be triggered? If so, I would rather generate an error message here, since the `exec` call below will fail anyway.

rszibele commented:
Hello @snoyberg, thank you for your code review, it has been very insightful. I have updated my pull request according to your suggestions. I have also added docker support, as it was the code path (`ExecOptsPlain`) which I mistakenly left out during the implementation.

I have re-tested everything and also tested it with docker enabled.

Wizek commented:
🎉 Hooray! Thanks guys!
